bin/install_protoc.sh

```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+if [ $# -ne 1 ]; then
+    echo $0: usage: install_protoc.sh path_to_install
+    echo ex: install_protoc.sh $GOPATH/src
+    exit 1
+fi
+
+# Check and Install protoc version 3.5.1
+protoc --help > /dev/null 2>&1
+if [[ $? -ne 0 || (`protoc --version` != *"libprotoc 3.5.1"*) ]]; then
+    cd $1
+    LOCAL_ARCH=`uname -m`
+    echo "Installing protobuf-all"
+    case "${OSTYPE}" in
+      darwin*)
+        curl -OL https://github.com/google/protobuf/releases/download/v3.5.1/protoc-3.5.1-osx-$LOCAL_ARCH.zip
+        unzip protoc-3.5.1-osx-$LOCAL_ARCH.zip -d protobuf-all
+        rm -f protoc-3.5.1-osx-$LOCAL_ARCH.zip ;;
+      linux*)
+        curl -OL https://github.com/google/protobuf/releases/download/v3.5.1/protoc-3.5.1-linux-$LOCAL_ARCH.zip
+        unzip protoc-3.5.1-linux-$LOCAL_ARCH.zip -d protobuf-all
+        rm -f protoc-3.5.1-linux-$LOCAL_ARCH.zip;;
+      msys)
+        curl -OL https://github.com/google/protobuf/releases/download/v3.5.1/protoc-3.5.1-win32.zip
+        unzip protoc-3.5.1-win32.zip -d protobuf-all
+        rm -f protoc-3.5.1-win32.zip ;;
+      *)
+        echo "unknown: ${OSTYPE}"
+        exit 1;;
+    esac
+    sudo mv protobuf-all/bin/* /usr/local/bin
```

Comment:
This doesn't look right to me. Please do not modify the developer's machine state.

Reply:
I do this only if developer does not already have protoc installed and developer is suggested to run depend.diff only if they have vendor changes. In that case I think this is a help to developer so that all the required binaries for go generate are installed for them, if they don't have it already...

Also, we do install stuff for things like linters: 
https://github.com/istio/istio/blob/master/bin/linters.sh#L29:7

Reply:
Modifying /usr/.../bin is a big no-no. In general, it is a big source of subtle issues for the developer, in addition to being a potential security problem. If nothing else, the assumption that the machine is being used for Istio development only is wrong. (It is quite likely that it is used to develop multiple projects which may have different versioned tooling) From this regard, go get has relatively reduced scope (which is easy to isolate) and does not require privileges. 

Re: linters We actually used to do docker based linters. (I am not sure why we're not doing it anymore). Api repo uses docker images as well. 

It is better to document the prerequisites, and loudly complain when they are not found.

Reply:
Thanks.. that makes sense.. how about adding a big red warning in red and going ahead and installing only if user says yes..

Is there a way to check that script is running in circle ci? So that we don't wait for user input when running this script later in circle ci?

Reply:
We can document the prerequisite here:
https://github.com/istio/istio/wiki/Dev-Guide

Also, why do we need protoc 3.5.1 explicitly? Do you need to peg the version number to get consistent diffs? 

A cleaner approach is probably to update all go:generates that makes direct calls to protoc to be updated to use one from a docker image. The API repo already uses a protoc in a docker image, so you can simply use it. To make things simple, you can create a "protoc.sh" in bin/, that acts just like regular protoc, but uses docker run internally (i.e. it simply passes its command-line flags to docker run). 
https://github.com/istio/api/blob/master/Makefile

Reply:
pinned protoc version is indeed needed to get a consistent diff.

Reply:
@ozevren : what is the version used here? https://github.com/istio/api/blob/master/Makefile#L14:32